File: sdks/python/src/scaleplane/transport.py
from __future__ import annotations

from typing import Any

import httpx

from scaleplane.auth import AuthProvider
from scaleplane.config import ClientConfig
from scaleplane.errors import ApiError


class HttpTransport:
    """Closed HTTP layer — resources call this; do not extend per feature."""

    def __init__(self, config: ClientConfig, auth: AuthProvider) -> None:
        self._config = config
        self._auth = auth

    def request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json: Any = None,
        accept_statuses: set[int] | None = None,
    ) -> Any:
        url = f"{self._config.base_url}{path}"
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            **self._auth.headers(),
        }
        with httpx.Client(timeout=self._config.timeout) as client:
            response = client.request(method, url, headers=headers, params=params, json=json)

        ok = response.status_code < 400 or (
            accept_statuses is not None and response.status_code in accept_statuses
        )
        if not ok:
            detail = _extract_detail(response)
            raise ApiError(response.status_code, detail)

        if response.status_code == 204 or not response.content:
            return None
        return response.json()


def _extract_detail(response: httpx.Response) -> str:
    try:
        body = response.json()
        if isinstance(body, dict) and "detail" in body:
            detail = body["detail"]
            return detail if isinstance(detail, str) else str(detail)
    except Exception:
        pass
    return response.text or response.reason_phrase

File: sdks/typescript/src/transport.ts
import type { AuthProvider } from './auth.js'
import type { ClientConfig } from './config.js'
import { ApiError } from './errors.js'

export type JsonValue = unknown

export class HttpTransport {
  constructor(
    private readonly config: ClientConfig,
    private readonly auth: AuthProvider,
  ) {}

  async request<T = JsonValue>(
    method: string,
    path: string,
    options: {
      params?: Record<string, string | number | undefined>
      body?: unknown
      acceptStatuses?: number[]
    } = {},
  ): Promise<T> {
    const url = new URL(`${this.config.baseUrl}${path}`)
    if (options.params) {
      for (const [key, value] of Object.entries(options.params)) {
        if (value !== undefined) {
          url.searchParams.set(key, String(value))
        }
      }
    }

    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.config.timeoutMs)
    try {
      const response = await fetch(url, {
        method,
        headers: {
          'Content-Type': 'application/json',
          Accept: 'application/json',
          ...this.auth.headers(),
        },
        body: options.body === undefined ? undefined : JSON.stringify(options.body),
        signal: controller.signal,
      })

      const accepted =
        response.status < 400 ||
        (options.acceptStatuses?.includes(response.status) ?? false)

      if (!accepted) {
        throw new ApiError(response.status, await extractDetail(response))
      }

      if (response.status === 204) {
        return undefined as T
      }
      const text = await response.text()
      if (!text) {
        return undefined as T
      }
      return JSON.parse(text) as T
    } finally {
      clearTimeout(timer)
    }
  }
}

async function extractDetail(response: Response): Promise<string> {
  try {
    const body = (await response.json()) as { detail?: unknown }
    if (body && typeof body.detail === 'string') {
      return body.detail
    }
    if (body?.detail !== undefined) {
      return String(body.detail)
    }
  } catch {
    // ignore
  }
  return response.statusText || 'Request failed'
}
